exposed-dao/src/main/kotlin/org/jetbrains/exposed/dao/Entity.kt

```diff
@@ -111,16 +111,20 @@ open class Entity<ID:Comparable<ID>>(val id: EntityID<ID>) {
         klass.invalidateEntityInCache(o)
         val currentValue = _readValues?.getOrNull(this)
         if (writeValues.containsKey(this as Column<out Any?>) || currentValue != value) {
+            val entityCache = TransactionManager.current().entityCache
             if (referee != null) {
-                val entityCache = TransactionManager.current().entityCache
                 if (value is EntityID<*> && value.table == referee!!.table) value.value // flush
 
                 listOfNotNull<Any>(value, currentValue).forEach {
-                    entityCache.referrers[it]?.remove(this)
+                    entityCache.referrers[this]?.remove(it)
                 }
                 entityCache.removeTablesReferrers(listOf(referee!!.table))
             }
             writeValues[this as Column<Any?>] = value
+            // TODO: Can this be simplified?
+            if (o.id._value?.let { entityCache.data[table]?.contains(it) } == true) {
```

Comment:
Presence of new entities in `inserts` during `init` is inconsistent, but they are only ever added to `data` after they are flushed.